core/bindings.rs

```diff
@@ -920,3 +924,102 @@ fn throw_type_error(scope: &mut v8::HandleScope, message: impl AsRef<str>) {
   let exception = v8::Exception::type_error(scope, message);
   scope.throw_exception(exception);
 }
+
+fn heap_stats(
+  scope: &mut v8::HandleScope,
+  _args: v8::FunctionCallbackArguments,
+  mut rv: v8::ReturnValue,
+) {
+  fn set_prop(
+    scope: &mut v8::HandleScope,
+    obj: v8::Local<v8::Object>,
+    name: &'static str,
+    value: usize,
+  ) {
+    let key = v8::String::new(scope, name).unwrap();
+    let val = v8::Number::new(scope, value as f64);
+    obj.set(scope, key.into(), val.into());
+  }
+
+  let s = get_heap_stats(scope);
+
+  // TODO: use serde for this once we have serde_v8
+  let obj = v8::Object::new(scope);
+  set_prop(scope, obj, "total_heap_size", s.total_heap_size);
+  set_prop(
+    scope,
+    obj,
+    "total_heap_size_executable",
+    s.total_heap_size_executable,
+  );
+  set_prop(scope, obj, "total_physical_size", s.total_physical_size);
+  set_prop(scope, obj, "total_available_size", s.total_available_size);
+  set_prop(
+    scope,
+    obj,
+    "total_global_handles_size",
+    s.total_global_handles_size,
+  );
+  set_prop(
+    scope,
+    obj,
+    "used_global_handles_size",
+    s.used_global_handles_size,
+  );
+  set_prop(scope, obj, "used_heap_size", s.used_heap_size);
+  set_prop(scope, obj, "heap_size_limit", s.heap_size_limit);
+  set_prop(scope, obj, "malloced_memory", s.malloced_memory);
+  set_prop(scope, obj, "external_memory", s.external_memory);
+  set_prop(scope, obj, "peak_malloced_memory", s.peak_malloced_memory);
+  set_prop(
+    scope,
+    obj,
+    "number_of_native_contexts",
+    s.number_of_native_contexts,
+  );
+  set_prop(
+    scope,
+    obj,
+    "number_of_detached_contexts",
+    s.number_of_detached_contexts,
+  );
+
+  rv.set(obj.into());
+}
+
+// HeapStats stores values from a isolate.get_heap_statistics() call
+struct HeapStats {
```

Comment:
I assume you're wrapping `v8::HeapStatistics` because that makes it easier to serialize in the future?

Reply:
Exactly, with `serde_v8` (and the new op-ABI) all the `set_prop()` calls would become a one-liner `to_v8()` and by keeping it in a struct we can have serde handle the camelCase-ification.

There are other core binding funcs that will be simplified with `serde_v8`, e.g: `get_proxy_details` and `get_promise_details` conceptually return tuples, which `serde_v8` can serialize to JS arrays (removing a fair amount of boilerplate in each function)

Reply:
I could probably add some simple alloc-dealloc tests (of large arrays) to ensure they are correctly reflected in the metrics.

Anything else that warrants testing in your opinion ?

Reply:
@AaronO such test would be sufficient